File: go/kth_smallest_ele_in_bst.go
package main

import (
	"leet_code/tree"
)

// 230. Kth Smallest Element in a BST
// Given the root of a binary search tree, and an integer k, return the kth (1-indexed) smallest element in the tree.
// Input: root = [3,1,4,null,2], k = 1
// Output: 1
// 230. 二叉搜索树中第K小的元素
// 给定一个二叉搜索树的根节点 root ，和一个整数 k ，请你设计一个算法查找其中第 k 个最小元素（从 1 开始计数）。
func main()  {
	node := node()
	kthSmallest(&node,5)
}

// 1.二叉搜索树中序遍历是升序，升序排序，然后找第k个元素
// 时间复杂度 O(N)
func kthSmallest(root *tree.TreeNode, k int) int {
	var ret,rank int
	traverse1(root,k,&ret,&rank)
	return ret
}

func traverse1(root *tree.TreeNode,k int,ret,rank *int)  {
	if root == nil {
		return
	}
	traverse1(root.Left,k,ret,rank)
	*rank++
	if k == *rank {
		*ret = root.Val
		return
	}
	traverse1(root.Right,k,ret,rank)
	return
}

func node() tree.TreeNode {
	node := tree.TreeNode{
		Val: 5,
		Left: &tree.TreeNode{
			Val: 3,
			Left: &tree.TreeNode{
				Val: 2,
				Left: &tree.TreeNode{
					Val: 1,
				},
			},
			Right: &tree.TreeNode{
				Val: 4,
			},
		},
		Right: &tree.TreeNode{
			Val: 6,
		},
	}
	return node
}

File: php/kth_smallest_ele_in_bst.php
<?php


//  Definition for a binary tree node.
//230. Kth Smallest Element in a BST
//Given the root of a binary search tree, and an integer k, return the kth (1-indexed) smallest element in the tree.
//Input: root = [3,1,4,null,2], k = 1
//Output: 1
class TreeNode {
  public $val = null;
  public $left = null;
  public $right = null;
  function __construct($val = 0, $left = null, $right = null) {
      $this->val = $val;
      $this->left = $left;
      $this->right = $right;
  }
}

// 1.二叉搜索树中序遍历结果是升序，因此使用递归，时间复杂度O(N)

class Solution {

    /**
     * @param TreeNode $root
     * @param Integer $k
     * @return Integer
     */
    function kthSmallest($root, $k) {
        $rank = $ret = 0;
        $this->traverse($root,$k,$rank,$ret);
        var_dump($ret);
    }

    function traverse($root,$k,&$rank,&$ret){
        if(empty($root)){
            return;
        }
        // 中序遍历
        $this->traverse($root->left,$k,$rank,$ret);
        $rank++;
        if($rank == $k){
            $ret = $root->val;
            return;
        }
        $this->traverse($root->right,$k,$rank,$ret);
        return;
    }

    function node(){
        $node = new TreeNode(5);
        $node->left = new TreeNode(3);
        $node->left->left = new TreeNode(2);
        $node->left->left->left = new TreeNode(1);
        $node->right = new TreeNode(6);
        $node->left->right = new TreeNode(4);
        return $node;
    }
}
$obj = new Solution();
$node = $obj->node();
$obj->kthSmallest($node,5);

